File: iOS/DataPool/Sources/DataPool/Infrastructure/PoolMemberRemoteRepository.swift
import Core

class PoolMemberRemoteRepository: PoolMemberRepository {
    private let poolMemberRemoteDataSource: PoolMemberRemoteDataSource

    private let networkErrorHandler: NetworkErrorHandler

    init(
        poolMemberRemoteDataSource: PoolMemberRemoteDataSource,
        networkErrorHandler: NetworkErrorHandler
    ) {
        self.poolMemberRemoteDataSource = poolMemberRemoteDataSource
        self.networkErrorHandler = networkErrorHandler
    }

    func getPoolMembers(
        poolId: String,
        next: String?
    ) async -> Result<CursorPage<PoolMember>, Error> {
        return await networkErrorHandler.handle {
            let page = try await poolMemberRemoteDataSource.getPoolMembers(
                poolId: poolId,
                next: next
            )
            return page.map { poolMemberResponse in
                poolMemberResponse.toPoolMember()
            }
        }
    }

    func removeGambler(
        poolId: String,
        gamblerId: String
    ) async -> Result<Void, Error> {
        return await networkErrorHandler.handle {
            try await poolMemberRemoteDataSource.removeGambler(
                poolId: poolId,
                gamblerId: gamblerId
            )
        }
    }
}

File: Android/data/pool/src/main/java/com/felipearpa/tyche/data/pool/infrastructure/PoolMemberRemoteRepository.kt
package com.felipearpa.tyche.data.pool.infrastructure

import com.felipearpa.network.NetworkExceptionHandler
import com.felipearpa.tyche.core.paging.CursorPage
import com.felipearpa.tyche.core.paging.map
import com.felipearpa.tyche.data.pool.domain.PoolMember
import com.felipearpa.tyche.data.pool.domain.PoolMemberDataSource
import com.felipearpa.tyche.data.pool.domain.PoolMemberRepository
import com.felipearpa.tyche.data.pool.domain.toPoolMember

internal class PoolMemberRemoteRepository(
    private val poolMemberDataSource: PoolMemberDataSource,
    private val networkExceptionHandler: NetworkExceptionHandler,
) : PoolMemberRepository {
    override suspend fun getPoolMembers(
        poolId: String,
        next: String?,
    ): Result<CursorPage<PoolMember>> {
        return networkExceptionHandler.handle {
            poolMemberDataSource.getPoolMembers(poolId = poolId, next = next)
                .map { poolMemberResponse -> poolMemberResponse.toPoolMember() }
        }
    }

    override suspend fun removeGambler(poolId: String, gamblerId: String): Result<Unit> {
        return networkExceptionHandler.handle {
            poolMemberDataSource.removeGambler(poolId = poolId, gamblerId = gamblerId)
        }
    }
}
